File: apps/backend-java/src/main/java/com/luban/backend/service/CollectionService.java
package com.luban.backend.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.luban.backend.dto.CollectionItemResponse;
import com.luban.backend.dto.CollectionResponse;
import com.luban.backend.entity.Collection;
import com.luban.backend.entity.CollectionItem;
import com.luban.backend.exception.BusinessException;
import com.luban.backend.mapper.CollectionMapper;
import com.luban.backend.mapper.SiteMapper;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * V2-T7 Collection CMS 服务。
 * Collection + CollectionItem CRUD；siteId tenant guard；name 冲突检测。
 */
@Service
public class CollectionService {

    private final CollectionMapper collectionMapper;
    private final SiteMapper siteMapper;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public CollectionService(CollectionMapper collectionMapper, SiteMapper siteMapper) {
        this.collectionMapper = collectionMapper;
        this.siteMapper = siteMapper;
    }

    // === Collection ===

    public List<CollectionResponse> list(String siteId) {
        if (siteMapper.getById(siteId) == null) throw BusinessException.siteNotFound();
        return collectionMapper.listBySiteId(siteId).stream()
            .map(CollectionResponse::fromEntity).collect(Collectors.toList());
    }

    public CollectionResponse get(String siteId, String id) {
        Collection c = collectionMapper.getByIdAndSiteId(id, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        return CollectionResponse.fromEntity(c);
    }

    public CollectionResponse create(String siteId, String name, JsonNode fieldSchema, String status) {
        if (siteMapper.getById(siteId) == null) throw BusinessException.siteNotFound();
        if (status == null || status.isBlank()) status = "active";
        Collection c = new Collection();
        c.setId(UUID.randomUUID().toString());
        c.setSiteId(siteId);
        c.setName(name);
        c.setFieldSchemaJson(jsonToString(fieldSchema));
        c.setStatus(status);
        Instant now = Instant.now();
        c.setCreatedAt(now);
        c.setUpdatedAt(now);
        try {
            collectionMapper.insert(c);
        } catch (DataIntegrityViolationException e) {
            if (e.getMessage() != null && e.getMessage().contains("Duplicate")) {
                throw BusinessException.collectionNameConflict();
            }
            throw e;
        }
        return CollectionResponse.fromEntity(c);
    }

    public CollectionResponse update(String siteId, String id, String name, JsonNode fieldSchema, String status) {
        Collection c = collectionMapper.getByIdAndSiteId(id, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        c.setName(name);
        if (fieldSchema != null) c.setFieldSchemaJson(jsonToString(fieldSchema));
        if (status != null) c.setStatus(status);
        c.setUpdatedAt(Instant.now());
        try {
            int n = collectionMapper.update(c);
            if (n == 0) throw BusinessException.collectionNotFound();
        } catch (DataIntegrityViolationException e) {
            if (e.getMessage() != null && e.getMessage().contains("Duplicate")) {
                throw BusinessException.collectionNameConflict();
            }
            throw e;
        }
        return CollectionResponse.fromEntity(c);
    }

    public void delete(String siteId, String id) {
        int n = collectionMapper.deleteByIdAndSiteId(id, siteId);
        if (n == 0) throw BusinessException.collectionNotFound();
        // collection_items 由 FK ON DELETE CASCADE 自动清理
    }

    // === CollectionItem ===

    public List<CollectionItemResponse> listItems(String siteId, String collectionId) {
        // 校验 collection 属于该 site
        Collection c = collectionMapper.getByIdAndSiteId(collectionId, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        return collectionMapper.listItemsByCollectionId(collectionId).stream()
            .map(CollectionItemResponse::fromEntity).collect(Collectors.toList());
    }

    public CollectionItemResponse getItem(String siteId, String collectionId, String itemId) {
        Collection c = collectionMapper.getByIdAndSiteId(collectionId, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        CollectionItem it = collectionMapper.getItemByIdAndCollectionId(itemId, collectionId);
        if (it == null) throw BusinessException.collectionItemNotFound();
        return CollectionItemResponse.fromEntity(it);
    }

    public CollectionItemResponse createItem(String siteId, String collectionId, JsonNode data, String status) {
        Collection c = collectionMapper.getByIdAndSiteId(collectionId, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        if (status == null || status.isBlank()) status = "active";
        CollectionItem it = new CollectionItem();
        it.setId(UUID.randomUUID().toString());
        it.setCollectionId(collectionId);
        it.setDataJson(jsonToString(data));
        it.setStatus(status);
        Instant now = Instant.now();
        it.setCreatedAt(now);
        it.setUpdatedAt(now);
        collectionMapper.insertItem(it);
        return CollectionItemResponse.fromEntity(it);
    }

    public CollectionItemResponse updateItem(String siteId, String collectionId, String itemId, JsonNode data, String status) {
        Collection c = collectionMapper.getByIdAndSiteId(collectionId, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        CollectionItem it = collectionMapper.getItemByIdAndCollectionId(itemId, collectionId);
        if (it == null) throw BusinessException.collectionItemNotFound();
        if (data != null) it.setDataJson(jsonToString(data));
        if (status != null) it.setStatus(status);
        it.setUpdatedAt(Instant.now());
        int n = collectionMapper.updateItem(it);
        if (n == 0) throw BusinessException.collectionItemNotFound();
        return CollectionItemResponse.fromEntity(it);
    }

    public void deleteItem(String siteId, String collectionId, String itemId) {
        Collection c = collectionMapper.getByIdAndSiteId(collectionId, siteId);
        if (c == null) throw BusinessException.collectionNotFound();
        int n = collectionMapper.deleteItemByIdAndCollectionId(itemId, collectionId);
        if (n == 0) throw BusinessException.collectionItemNotFound();
    }

    private String jsonToString(JsonNode node) {
        if (node == null) return "{}";
        try {
            return objectMapper.writeValueAsString(node);
        } catch (Exception e) {
            return "{}";
        }
    }
}

File: apps/backend-go/internal/service/collection_service.go
package service

import (
	"context"
	"encoding/json"

	"github.com/google/uuid"
	"github.com/xiaoshuai1024/luban-backend-go/internal/model"
	"github.com/xiaoshuai1024/luban-backend-go/internal/repository"
)

// V2-T7 CollectionService — 对齐 Java CollectionService（CRUD + siteId guard）。
type CollectionService struct {
	repo     *repository.CollectionRepository
	siteRepo *repository.SiteRepository
}

func NewCollectionService(repo *repository.CollectionRepository, siteRepo *repository.SiteRepository) *CollectionService {
	return &CollectionService{repo: repo, siteRepo: siteRepo}
}

// === Collection ===

func (s *CollectionService) List(ctx context.Context, siteID string) ([]*model.Collection, error) {
	if _, err := s.siteRepo.Get(ctx, siteID); err != nil {
		return nil, repository.ErrSiteNotFound
	}
	return s.repo.ListBySite(ctx, siteID)
}

func (s *CollectionService) Get(ctx context.Context, siteID, id string) (*model.Collection, error) {
	return s.repo.Get(ctx, siteID, id)
}

func (s *CollectionService) Create(ctx context.Context, siteID, name string, fieldSchema json.RawMessage, status string) (*model.Collection, error) {
	if _, err := s.siteRepo.Get(ctx, siteID); err != nil {
		return nil, repository.ErrSiteNotFound
	}
	if status == "" {
		status = "active"
	}
	if len(fieldSchema) == 0 {
		fieldSchema = json.RawMessage(`{}`)
	}
	c := &model.Collection{
		ID:          uuid.NewString(),
		SiteID:      siteID,
		Name:        name,
		FieldSchema: fieldSchema,
		Status:      status,
	}
	if err := s.repo.Create(ctx, c); err != nil {
		return nil, err
	}
	return c, nil
}

func (s *CollectionService) Update(ctx context.Context, c *model.Collection) error {
	return s.repo.Update(ctx, c)
}

func (s *CollectionService) Delete(ctx context.Context, siteID, id string) error {
	return s.repo.Delete(ctx, siteID, id)
}

// === CollectionItem ===

func (s *CollectionService) ListItems(ctx context.Context, siteID, collectionID string) ([]*model.CollectionItem, error) {
	if _, err := s.repo.Get(ctx, siteID, collectionID); err != nil {
		return nil, err
	}
	return s.repo.ListItems(ctx, collectionID)
}

func (s *CollectionService) GetItem(ctx context.Context, siteID, collectionID, itemID string) (*model.CollectionItem, error) {
	if _, err := s.repo.Get(ctx, siteID, collectionID); err != nil {
		return nil, err
	}
	return s.repo.GetItem(ctx, collectionID, itemID)
}

func (s *CollectionService) CreateItem(ctx context.Context, siteID, collectionID string, data json.RawMessage, status string) (*model.CollectionItem, error) {
	if _, err := s.repo.Get(ctx, siteID, collectionID); err != nil {
		return nil, err
	}
	if status == "" {
		status = "active"
	}
	if len(data) == 0 {
		data = json.RawMessage(`{}`)
	}
	it := &model.CollectionItem{
		ID:           uuid.NewString(),
		CollectionID: collectionID,
		Data:         data,
		Status:       status,
	}
	if err := s.repo.CreateItem(ctx, it); err != nil {
		return nil, err
	}
	return it, nil
}

func (s *CollectionService) UpdateItem(ctx context.Context, it *model.CollectionItem) error {
	return s.repo.UpdateItem(ctx, it)
}

func (s *CollectionService) DeleteItem(ctx context.Context, siteID, collectionID, itemID string) error {
	if _, err := s.repo.Get(ctx, siteID, collectionID); err != nil {
		return err
	}
	return s.repo.DeleteItem(ctx, collectionID, itemID)
}
